// pdf-to-png-converter/out/cli.js
#!/usr/bin/env node
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HELP_TEXT = void 0;
exports.parseBoolean = parseBoolean;
exports.parseNumberList = parseNumberList;
exports.getVersion = getVersion;
exports.run = run;
const node_util_1 = require("node:util");
const node_path_1 = __importDefault(require("node:path"));
const node_fs_1 = __importDefault(require("node:fs"));
const pdfToPng_js_1 = require("./pdfToPng.js");
/**
 * Help text shown for `--help` and on invalid usage.
 * Exported so it can be asserted in tests without running the full CLI pipeline.
 */
exports.HELP_TEXT = `Usage: pdf-to-png-converter <pdf-file-path> [options]

Options:
  --output-folder <dir>             Folder path where PNG files will be written
  --viewport-scale <number>         Scale factor applied to each page viewport
  --use-system-fonts                Attempt to use fonts installed on the host system
  --disable-font-face <true|false>  Do not load embedded fonts (true/false)
  --enable-xfa <true|false>         Process XFA form data (true/false)
  --pdf-file-password <pwd>         Password for encrypted PDFs
  --pages-to-process <n,m,...>      Comma-separated list of 1-based page numbers
  --verbosity-level <number>        pdfjs verbosity level (0=errors, 1=warnings, 5=infos)
  --return-metadata-only            Return page metadata without rendering images
  --return-page-content             Retain rendered PNG buffers in results (default: false)
  --process-pages-in-parallel       Process pages concurrently
  --concurrency-limit <number>      Maximum number of pages rendered simultaneously
  --silent                          Suppress output unless there is an error
  --version                         Show version
  --help                            Show this help message`;
/**
 * Shared `parseArgs` option schema.
 * Declared `as const` so TypeScript infers precise `'string'` / `'boolean'` literal types
 * for every flag and propagates them through the `parseArgs` return type.
 */
const CLI_OPTIONS = {
    'output-folder': { type: 'string' },
    'viewport-scale': { type: 'string' },
    'use-system-fonts': { type: 'boolean' },
    'disable-font-face': { type: 'string' },
    'enable-xfa': { type: 'string' },
    'pdf-file-password': { type: 'string' },
    'pages-to-process': { type: 'string' },
    'verbosity-level': { type: 'string' },
    'return-metadata-only': { type: 'boolean' },
    'return-page-content': { type: 'boolean' },
    'process-pages-in-parallel': { type: 'boolean' },
    'concurrency-limit': { type: 'string' },
    silent: { type: 'boolean' },
    version: { type: 'boolean' },
    help: { type: 'boolean' },
};
/**
 * Parses a CLI string value as a boolean.
 *
 * - `'true'` / `'1'` → `true`
 * - `'false'` / `'0'` → `false`
 * - `undefined` → `undefined` (flag not provided)
 *
 * @throws {Error} When the value is not a recognised boolean string.
 */
function parseBoolean(val) {
    if (val === undefined)
        return undefined;
    if (val === 'true' || val === '1')
        return true;
    if (val === 'false' || val === '0')
        return false;
    throw new Error(`Invalid boolean value: "${val}". Expected true|false|1|0.`);
}
/**
 * Parses a comma-separated string of integers into a `number[]`.
 *
 * Returns `undefined` when `val` is `undefined` (flag not provided).
 *
 * @throws {Error} When any token in the list is not a valid integer.
 */
function parseNumberList(val) {
    if (val === undefined)
        return undefined;
    return val.split(',').map((token) => {
        const trimmed = token.trim();
        if (trimmed === '')
            throw new Error('Invalid integer in list: empty value.');
        const parsed = Number(trimmed);
        if (!Number.isInteger(parsed))
            throw new Error(`Invalid integer in list: "${trimmed}".`);
        return parsed;
    });
}
/**
 * Reads the package version from the adjacent `package.json`.
 * Falls back to `'Unknown'` on any I/O or parse error so that `--version`
 * always returns a printable string.
 */
function getVersion() {
    try {
        const pkgPath = node_path_1.default.resolve(__dirname, '../package.json');
        const pkgInfo = JSON.parse(node_fs_1.default.readFileSync(pkgPath, 'utf-8'));
        return pkgInfo.version ?? 'Unknown';
    }
    catch {
        return 'Unknown';
    }
}
/**
 * Main CLI entry point.
 *
 * Parses `process.argv`, validates all options up-front with actionable error messages,
 * and delegates to {@link pdfToPng}. Exported so it can be unit-tested without
 * spawning a child process.
 */
async function run() {
    // parseArgs throws on unknown flags (strict mode is on by default).
    // Wrap in an IIFE that returns null on failure so the outer scope
    // can use a const binding and guard cleanly when process.exit is mocked.
    const parseResult = (() => {
        try {
            return (0, node_util_1.parseArgs)({ options: CLI_OPTIONS, allowPositionals: true });
        }
        catch (err) {
            /* v8 ignore next */
            console.error(err instanceof Error ? err.message : String(err));
            console.error(exports.HELP_TEXT);
            process.exit(1);
            return null; // only reached when process.exit is mocked (e.g. unit tests)
        }
    })();
    // Guard: if process.exit was mocked and parseArgs failed, return early.
    if (!parseResult)
        return;
    const { values, positionals } = parseResult;
    if (values.help) {
        console.log(exports.HELP_TEXT);
        process.exit(0);
        return;
    }
    if (values.version) {
        console.log(`v${getVersion()}`);
        process.exit(0);
        return;
    }
    const pdfFilePath = positionals[0];
    if (!pdfFilePath) {
        console.error('Error: <pdf-file-path> is required.');
        console.error(exports.HELP_TEXT);
        process.exit(1);
        return;
    }
    // ── Validate numeric options up-front for actionable error messages ───────
    let viewportScale;
    if (values['viewport-scale'] !== undefined) {
        viewportScale = Number(values['viewport-scale']);
        if (!Number.isFinite(viewportScale)) {
            console.error('Error: --viewport-scale must be a valid number.');
            process.exit(1);
            return;
        }
    }
    let verbosityLevel;
    if (values['verbosity-level'] !== undefined) {
        verbosityLevel = Number(values['verbosity-level']);
        if (!Number.isInteger(verbosityLevel)) {
            console.error('Error: --verbosity-level must be a valid integer.');
            process.exit(1);
            return;
        }
    }
    let concurrencyLimit;
    if (values['concurrency-limit'] !== undefined) {
        concurrencyLimit = Number(values['concurrency-limit']);
        if (!Number.isInteger(concurrencyLimit)) {
            console.error('Error: --concurrency-limit must be a valid integer.');
            process.exit(1);
            return;
        }
    }
    // ── Validate boolean / list options (helpers throw on bad input) ──────────
    let disableFontFace;
    let enableXfa;
    let pagesToProcess;
    try {
        disableFontFace = parseBoolean(values['disable-font-face']);
        enableXfa = parseBoolean(values['enable-xfa']);
        pagesToProcess = parseNumberList(values['pages-to-process']);
    }
    catch (err) {
        /* v8 ignore next */
        console.error(`Error: ${err instanceof Error ? err.message : String(err)}`);
        process.exit(1);
        return;
    }
    // ── Build options and invoke pdfToPng ─────────────────────────────────────
    /** Logs to stdout unless `--silent` is set. */
    const log = (...msgs) => {
        if (!values.silent)
            console.log(...msgs);
    };
    // All PdfToPngOptions properties are set explicitly (to undefined when not provided)
    // so the object passed to pdfToPng is fully predictable and easy to assert in tests.
    const options = {
        outputFolder: values['output-folder'],
        viewportScale,
        useSystemFonts: values['use-system-fonts'],
        disableFontFace,
        enableXfa,
        pdfFilePassword: values['pdf-file-password'],
        pagesToProcess,
        verbosityLevel,
        returnMetadataOnly: values['return-metadata-only'],
        returnPageContent: values['return-page-content'] ?? false,
        processPagesInParallel: values['process-pages-in-parallel'],
        concurrencyLimit,
    };
    log(`Processing PDF: ${pdfFilePath}`);
    if (options.outputFolder) {
        log(`Output folder: ${options.outputFolder}`);
    }
    try {
        const results = await (0, pdfToPng_js_1.pdfToPng)(pdfFilePath, options);
        if (values['return-metadata-only']) {
            log('Metadata extraction complete:');
            log(JSON.stringify(results, null, 2));
        }
        else {
            log(`Successfully processed ${results.length} page(s).`);
        }
    }
    catch (err) {
        console.error('Error:');
        console.error(err instanceof Error ? err.message : String(err));
        process.exit(1);
    }
}
/* v8 ignore next 6 */
try {
    if (node_fs_1.default.realpathSync(process.argv[1]) === node_fs_1.default.realpathSync(__filename)) {
        void run();
    }
}
catch {
    // realpathSync can fail (e.g. path does not exist) — skip auto-execution
}
