// pdf-to-png-converter/out/const.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DOCUMENT_INIT_PARAMS_DEFAULTS = exports.STANDARD_FONTS_RELATIVE_URL = exports.CMAP_RELATIVE_URL = exports.PDF_TO_PNG_OPTIONS_DEFAULTS = exports.MAX_CANVAS_PIXELS = exports.MAX_VIEWPORT_SCALE = void 0;
/**
 * Maximum allowed value for `viewportScale`. Values above this limit would produce canvases
 * so large (an A4 page at scale 100 already yields ~5×10⁹ pixels) that they risk OOM crashes
 * before the pixel-count guard in `processPdfPage` can fire.
 */
exports.MAX_VIEWPORT_SCALE = 100;
/**
 * Maximum canvas area in pixels. At 4 bytes per pixel, 100 MP ≈ 400 MB of raw bitmap memory.
 * Any page whose viewport exceeds this limit will be rejected before canvas allocation.
 */
exports.MAX_CANVAS_PIXELS = 100_000_000;
/**
 * Default values applied to `PdfToPngOptions` fields that are not explicitly set by the caller.
 * These are also used as the source of truth for documented defaults in JSDoc comments on the type.
 */
exports.PDF_TO_PNG_OPTIONS_DEFAULTS = {
    viewportScale: 1,
    disableFontFace: true,
    useSystemFonts: false,
    enableXfa: true,
    /** Used as the output filename stem when the PDF is supplied as a buffer rather than a file path. */
    outputFileMask: 'buffer',
    pdfFilePassword: undefined,
    concurrencyLimit: 4,
};
/**
 * Relative paths to the pdfjs-dist asset directories.
 * Stored as raw strings so they can be resolved against `process.cwd()` at call time
 * (inside `propsToPdfDocInitParams`) rather than at module-load time. This ensures
 * applications that call `process.chdir()` after importing the library still get
 * correct paths.
 */
exports.CMAP_RELATIVE_URL = './node_modules/pdfjs-dist/cmaps/';
exports.STANDARD_FONTS_RELATIVE_URL = './node_modules/pdfjs-dist/standard_fonts/';
/**
 * Default pdfjs `DocumentInitParameters` used when initialising a PDF document.
 * - `cMapUrl` / `cMapPacked`: point to the pre-packed character maps bundled with `pdfjs-dist`,
 *    required for rendering CJK and other non-Latin PDFs correctly.
 * - `standardFontDataUrl`: path to the standard Type 1 / TrueType fonts bundled with `pdfjs-dist`,
 *    used as fallbacks when a PDF does not embed its fonts.
 *
 * Note: these values are raw relative paths. They are resolved to absolute paths at call time
 * by `propsToPdfDocInitParams` via `normalizePath`.
 */
exports.DOCUMENT_INIT_PARAMS_DEFAULTS = {
    cMapUrl: exports.CMAP_RELATIVE_URL,
    cMapPacked: true,
    standardFontDataUrl: exports.STANDARD_FONTS_RELATIVE_URL,
};
// Test-only asset lists (STANDARD_FONTS, STANDARD_CMAPS) live in __tests__/test-data-constants.ts
